File: src-backend/models/transaction.py
from datetime import date
from typing import Optional

from sqlalchemy import Date, Enum, ForeignKey, Integer, Numeric, Text
from sqlalchemy.orm import Mapped, mapped_column, relationship

from core.database import Base
from enums.currency_enum import Currency
from enums.transaction_type_enum import TransactionType


class Transaction(Base):
    __tablename__ = "transactions"

    id: Mapped[int] = mapped_column(primary_key=True)
    account_id: Mapped[int] = mapped_column(ForeignKey("accounts.id"), nullable=False)
    category_id: Mapped[Optional[int]] = mapped_column(
        ForeignKey("categories.id"), nullable=True
    )
    type: Mapped[TransactionType] = mapped_column(Enum(TransactionType), nullable=False)
    amount: Mapped[float] = mapped_column(Numeric(18, 2), nullable=False)
    currency: Mapped[Currency] = mapped_column(Enum(Currency), nullable=False)
    date: Mapped[date] = mapped_column(Date, nullable=False)
    description: Mapped[Optional[str]] = mapped_column(Text, nullable=True)
    source_type: Mapped[Optional[str]] = mapped_column(Text, nullable=True)
    source_id: Mapped[Optional[int]] = mapped_column(Integer, nullable=True)

    account: Mapped["Account"] = relationship("Account", back_populates="transactions")
    category: Mapped[Optional["Category"]] = relationship(
        "Category", back_populates="transactions"
    )

File: packages/main/src/schemas/transaction.ts
import { Currency } from './enums.js'

export const TransactionType = {
  INCOME:   'income',
  EXPENSE:  'expense',
  TRANSFER: 'transfer',
} as const
export type TransactionType = typeof TransactionType[keyof typeof TransactionType]

export interface Transaction {
  id:             number
  account_id:     number
  category_id:    number | null
  transaction_type: TransactionType
  amount:         number
  currency:       Currency
  transacted_at:  string
  description:    string | null
}

export type NewTransaction = Omit<Transaction, 'id'>
